library/core/src/marker.rs

```diff
@@ -879,25 +879,35 @@ marker_impls! {
         {T: ?Sized} &mut T,
 }
 
-/// Types that can be safely moved after being pinned.
+/// Types that do not need to follow the rules of pinning.
 ///
-/// Rust itself has no notion of immovable types, and considers moves (e.g.,
-/// through assignment or [`mem::replace`]) to always be safe.
+/// For information on what "pinning" is, see the [`pin` module] documentation.
 ///
-/// The [`Pin`][Pin] type is used instead to prevent moves through the type
-/// system. Pointers `P<T>` wrapped in the [`Pin<P<T>>`][Pin] wrapper can't be
-/// moved out of. See the [`pin` module] documentation for more information on
-/// pinning.
+/// Implementing the `Unpin` trait for `T` lifts the restrictions of pinning off that type.
+/// This means that, if `T: Unpin`, it cannot be assumed that a value of type `T` will be bound
+/// by the invariants that pinning infers, *even* when "pinned" by a [`Pin<Ptr>`] pointing at it.
+/// When a value of type `T` is pointed at by a [`Pin<Ptr>`], [`Pin`] will not restrict access
+/// to the pointee value like it normally would, thus allowing the user to do anything that they
+/// normally could with a non-[`Pin`]-wrapped `Ptr` to that value.
 ///
-/// Implementing the `Unpin` trait for `T` lifts the restrictions of pinning off
-/// the type, which then allows moving `T` out of [`Pin<P<T>>`][Pin] with
-/// functions such as [`mem::replace`].
+/// The idea of this trait is to alleviate the reduced ergonomics of APIs that require the use
+/// of [`Pin`] for soundness for some types, but which also want to be used by other types that
+/// don't care about pinning. The prime example of such an API is [`Future::poll`]. There are many
```

Comment:
I'm strongly persuaded that we should accept this PR, or at least some of it, based on the strength of this improvement in the docs alone. If we do find any reason to block part of this, or even quibble for any serious length of time, I would prefer that we ask for it to be split up. `Pin` and `Unpin` is a rare case where trying to describe it in terms of "positive license", instead of negative, is mostly confounding.

Reply:
While that is true of non-`unsafe` Rust, I do think that the moment `unsafe` Rust is involved (which may very well happen when working with the pinning invariants), then knowing what _cannot_ happen with a given (non-`Unpin`) type is what allows the `unsafe` code to assume certain things:

  - _e.g._, `unsafe` code can rely on _absence of `Send`_ implying _absence_ of unique access crossing-thread boundaries, and likewise for `Sync` and shared access. Similarly, `unsafe` code can rely on absence of `Unpin` to know that witnessing a `Pin`-wrapped pointer to such a type means that the pointee's (shallow) memory shall not be invalidated (deällocated or relocated) before it is dropped.

  - (this kind of relates to, _conceptually_, `Unpin` almost having deserved to be an `unsafe` trait)

Reply:
Such negative reasoning is particularly fragile when using unsafe. Everything unsafe does should be positively justified from the type's invariants, not negatively justified from what `Pin` won't do. `Pin` is just a library type, it won't do certain things because the invariants won't let it, but it's the invariants that really matter.

Reply:
It is fragile, but it is also necessary, and this is one of the key sources of confusions behind this type. The mere existence of Pin has made reasoning about unsafe code more confusing _even though Pin itself doesn't have language level magic_. The current docs make people incorrectly question their understanding of the overall safety model of Rust. That's not good.

We can and should do both: explain in both positive and negative terms. The docs need to be able to get on the train of having a vaguely correct mental model _before_ it tries to give them a precise one. Currently the docs seem to leave people stranded at the station.

Reply:
I'm very aware that pinning is a language feature. I wrote a [blog post](https://www.ralfj.de/blog/2018/04/05/a-formal-look-at-pinning.html) about this years ago.

However, the new language feature is the concept if "memory-address-sensitive values" (to use the terminology from this PR). `Pin` is still just a library type, albeit one that deeply interacts with that concept.

Reply:
> I'm very aware that pinning is a language feature

No, I know you know this. What I'm saying is that this is one of the key sources of confusion for people. I think negative license is actually relevant for Pin.

For Unpin, yes, I think that's trickier. I still think this PR is an improvement and we can revisit Unpin specifically after this lands.

Reply:
>  I think negative license is actually relevant for Pin.

We remain in disagreement here. I find negative license basically never a good way to explain or understand things, certainly not in this case.

I'm not at all opposed to explaining this as a consequence arising from the underlying mechanism. I'm just not happy with presenting this as what `Pin` *is*.

@danielhenrymantilla 
> e.g., unsafe code can rely on absence of Send implying absence of unique access crossing-thread boundaries,

This is just contraposition. Whenever "P implies Q" is true, then also "not Q implies not P" is true. But almost always, "P implies Q" is still the thing you want to explain, and view "not Q implies not P" just as a consequence. A consequence worth mentioning, but not the defining property.

I find it basically always more clear to state a concept without negations rather than with negations, when possible.

Reply:
> I'm not at all opposed to explaining this as a consequence arising from the underlying mechanism. I'm just not happy with presenting this as what `Pin` _is_.

I think that would be fine too but I think the current text _does not achieve this well at all_. I think turning this text into something that talks about what pin "is" and then immediately goes into this as a consequence is not a major edit for these docs. I think that's a tweak of the first few paragraphs of Pin.

I do think the Unpin docs could do with more rearranging.


Fundamentally, I have found the existing positive license to be extremely confusing to people trying to learn this. I think negative license is a bad way to define things from a spec point of view, but I think from a docs/explainer point of view it can work, and I think what we have here is miles better than what existed in the past.

(Personally my favorite explanation of pin is by talking about it like a type of typestate but I'm not sure that's the right tack for the docs here either)

Reply:
Incidentally, I think I just encountered a [case](https://github.com/rust-lang/rust/issues/49206#issuecomment-1880167930) where negative reasoning easily leads to wrong conclusions, and focusing on the positive properties that we actually have and need to prove provides clarity.

I am worried we'll draw ourselves into a corner with a negative wording in the documentation, because people will take that as the spec and they might draw wrong conclusions and then we get "interesting" soundness issues down the road, when their wrong interpretation of the actual spec, based on misleading docs, collides with the actual spec.

Reply:
Yeah; I think you're coming at this from your position of dealing with opsem stuff and while I am sympathetic to the danger of these docs being misinterpretable, that is the case for most of Rust's opsem already: the docs are not sufficient for this. I'm coming at this from the position of "the _first_ thing these docs should achieve is teaching people what these types are about. Laser precision about soundness can be secondary".

We should absolutely tighten up the language here to include that laser precision as well, and thanks for filing the followup for that.